File: android/app/src/main/java/ru/bgtu_voenmeh/zapara/data/GroupHomework.kt
package ru.bgtu_voenmeh.zapara.data

import kotlinx.coroutines.CancellationException

data class LocalHomeworkMark(val id: String, val done: Boolean)

data class GroupCopyMark(val homeworkId: String, val memberId: String, val completed: Boolean)

data class GroupHomeworkBook(val local: List<LocalHomeworkMark>, val copies: List<GroupCopyMark>)

data class HomeworkShareInput(val share: Boolean, val signedIn: Boolean, val communityId: String)

/** Subject and text from the open editor, the group box, and whether this save creates a task. */
data class HomeworkEditorShare(val subject: String, val text: String, val share: Boolean, val isNew: Boolean)

data class HomeworkShareOutcome(val stored: Boolean, val sent: Boolean, val note: String)

object GroupHomework {
    const val SIGN_IN = "Войдите в аккаунт, чтобы отправить домашку группе."
    const val LOCAL_ONLY = "Вы ещё не в группе. Домашка сохранена только на этом устройстве."
    const val FAILED = "На устройстве сохранено. Группе отправить не получилось."
    const val SHARED = "Домашка продублирована всей группе."

    /** Store the editor pair first. The group call receives that same pair and runs only after the local row exists. */
    suspend fun saveEditor(
        editor: HomeworkEditorShare,
        signedIn: Boolean,
        communityId: String,
        saveLocal: suspend (subject: String, text: String) -> Unit,
        send: suspend (subject: String, text: String) -> Unit,
    ): HomeworkShareOutcome {
        val subject = editor.subject.trim()
        val text = editor.text.trim()
        return save(subject, text, HomeworkShareInput(editor.share && editor.isNew, signedIn, communityId), saveLocal, send)
    }

    suspend fun save(
        subject: String,
        text: String,
        input: HomeworkShareInput,
        saveLocal: suspend (String, String) -> Unit,
        send: suspend (String, String) -> Unit,
    ): HomeworkShareOutcome {
        saveLocal(subject, text)
        if (!input.share) return HomeworkShareOutcome(true, false, "")
        if (!input.signedIn) return HomeworkShareOutcome(true, false, SIGN_IN)
        if (input.communityId.isBlank()) return HomeworkShareOutcome(true, false, LOCAL_ONLY)
        return try {
            send(subject, text)
            HomeworkShareOutcome(true, true, SHARED)
        } catch (e: CancellationException) {
            throw e
        } catch (_: Exception) {
            HomeworkShareOutcome(true, false, FAILED)
        }
    }

    fun complete(
        local: List<LocalHomeworkMark>,
        copies: List<GroupCopyMark>,
        actorId: String,
        homeworkId: String,
        completed: Boolean,
    ): GroupHomeworkBook {
        val actor = actorId.trim()
        val copyId = homeworkId.trim()
        return GroupHomeworkBook(
            local = local.map { it.copy() },
            copies = copies.map { row ->
                if (row.homeworkId == copyId && row.memberId == actor) row.copy(completed = completed) else row.copy()
            },
        )
    }
}

File: src/Vograph.Core/Services/GroupHomework.cs
namespace Vograph.Core.Services;

public sealed record HomeworkShareInput(bool Share, bool SignedIn, string? CommunityId);

/// <summary>What the homework editor holds when the person presses save: the subject and the text they wrote,
/// whether the group box is on, and whether this is a new task. An edit does not send another group copy.</summary>
public sealed record HomeworkEditorShare(string Subject, string Text, bool Share, bool IsNew);

public sealed record HomeworkShareOutcome(bool Stored, bool Sent, string Note);

public sealed record LocalHomeworkMark(string Id, bool Done);

public sealed record GroupCopyMark(string HomeworkId, string MemberId, bool Completed);

public sealed record GroupHomeworkBook(IReadOnlyList<LocalHomeworkMark> Local, IReadOnlyList<GroupCopyMark> Copies);

/// <summary>Local homework is stored first. A group copy is sent only when the box is on, the person is signed in,
/// and the timetable group already has a community. One member's «Сделано» touches only that member's copy.</summary>
public static class GroupHomework
{
    public const string SignInNote = "Войдите в аккаунт, чтобы отправить домашку группе.";
    public const string LocalOnlyNote = "Вы ещё не в группе. Домашка сохранена только на этом устройстве.";
    public const string FailedNote = "На устройстве сохранено. Группе отправить не получилось.";
    public const string SharedNote = "Домашка продублирована всей группе.";

    /// <summary>Store the editor's subject and text on the device first, then submit that same pair to the group
    /// only when the box is on, the task is new, the person is signed in, and a membership is present.</summary>
    public static Task<HomeworkShareOutcome> SaveEditorAsync(
        HomeworkEditorShare editor, bool signedIn, string? communityId, Func<string, string, Task> saveLocal, Func<string, string, Task> send)
    {
        ArgumentNullException.ThrowIfNull(editor);
        ArgumentNullException.ThrowIfNull(saveLocal);
        ArgumentNullException.ThrowIfNull(send);
        var subject = (editor.Subject ?? "").Trim();
        var text = (editor.Text ?? "").Trim();
        return SaveCoreAsync(subject, text, new(editor.Share && editor.IsNew, signedIn, communityId), saveLocal, send);
    }

    public static Task<HomeworkShareOutcome> SaveAsync(
        string subject, string text, HomeworkShareInput input, Func<string, string, Task> saveLocal, Func<string, string, Task> send)
    {
        ArgumentNullException.ThrowIfNull(saveLocal);
        ArgumentNullException.ThrowIfNull(send);
        return SaveCoreAsync(subject, text, input, saveLocal, send);
    }

    private static async Task<HomeworkShareOutcome> SaveCoreAsync(
        string subject, string text, HomeworkShareInput input, Func<string, string, Task> saveLocal, Func<string, string, Task> send)
    {
        ArgumentNullException.ThrowIfNull(input);
        await saveLocal(subject, text).ConfigureAwait(false);
        if (!input.Share) return new(true, false, "");
        if (!input.SignedIn) return new(true, false, SignInNote);
        if (string.IsNullOrWhiteSpace(input.CommunityId)) return new(true, false, LocalOnlyNote);
        try
        {
            await send(subject, text).ConfigureAwait(false);
            return new(true, true, SharedNote);
        }
        catch (OperationCanceledException) { throw; }
        catch (Exception)
        {
            return new(true, false, FailedNote);
        }
    }

    public static GroupHomeworkBook Complete(
        IReadOnlyList<LocalHomeworkMark> local, IReadOnlyList<GroupCopyMark> copies, string actorId, string homeworkId, bool completed)
    {
        ArgumentNullException.ThrowIfNull(local);
        ArgumentNullException.ThrowIfNull(copies);
        var actor = (actorId ?? "").Trim();
        var id = (homeworkId ?? "").Trim();
        return new(
            local.Select(row => row with { }).ToList(),
            copies.Select(row => row.HomeworkId == id && row.MemberId == actor ? row with { Completed = completed } : row with { }).ToList());
    }
}
